src/care/LoopFuser.h

```diff
@@ -733,37 +983,62 @@ void LoopFuser::registerFree(care::host_device_ptr<T> & array) {
 
 // Start recording
 #define FUSIBLE_LOOPS_START { \
-   auto __fuser__ = LoopFuser::getInstance(); \
-   __fuser__->start(); \
-   __fuser__->preserveOrder(false); \
-   __fuser__->setScan(false); \
+   static FusedActionsObserver * __phase_observer = new FusedActionsObserver(); \
+   for ( FusedActions *__fuser__ : {static_cast<FusedActions *> (LoopFuser::getInstance()),static_cast<FusedActions *>(__phase_observer)}) { \
+      __fuser__->startRecording(); \
+      __fuser__->preserveOrder(false); \
+      __fuser__->setScan(false); \
+   } \
+   FusedActionsObserver::activeObserver = __phase_observer; \
 }
 
 #define FUSIBLE_LOOPS_PRESERVE_ORDER_START { \
-   LoopFuser::getInstance()->start(); \
-   LoopFuser::getInstance()->preserveOrder(true); \
+   static FusedActionsObserver * __phase_observer = new FusedActionsObserver(); \
+   for ( FusedActions *__fuser__ : {static_cast<FusedActions *> (LoopFuser::getInstance()),static_cast<FusedActions *>(__phase_observer)}) { \
+      __fuser__->startRecording(); \
+      __fuser__->preserveOrder(true); \
+      __fuser__->setScan(false); \
+   } \
+   FusedActionsObserver::activeObserver = __phase_observer; \
 }
 
 // Execute, then stop recording
-#define FUSIBLE_LOOPS_STOP { \
-   LoopFuser::getInstance()->stop(); \
-   LoopFuser::getInstance()->flush(); \
+#define _FUSIBLE_LOOPS_STOP(ASYNC) { \
+   for ( FusedActions *__fuser__ : {static_cast<FusedActions *> (LoopFuser::getInstance()),static_cast<FusedActions *>(FusedActionsObserver::activeObserver)}) { \
+      __fuser__->stopRecording(); \
+      __fuser__->flushActions(ASYNC); \
+   } \
+   FusedActionsObserver::activeObserver = nullptr; \
 }
 
+// Execute, then stop recording
+#define FUSIBLE_LOOPS_STOP _FUSIBLE_LOOPS_STOP(false)
+
+// Execute asynchronously, then stop recording
+#define FUSIBLE_LOOPS_STOP_ASYNC _FUSIBLE_LOOPS_STOP(true)
+
+
 // frees
 #define FUSIBLE_FREE(A) LoopFuser::getInstance()->registerFree(A);
 
 #else // defined(CARE_DEBUG) || defined(__GPUCC__)
 
 // in opt, non cuda builds, never start recording
-#define FUSIBLE_LOOPS_START { \
-   auto __fuser__ = LoopFuser::getInstance(); \
-   __fuser__->setScan(false); \
-   __fuser__->setCountsToOffsetsScan(false); \
+#define FUSIBLE_LOOPS_START \
+{ \
+   static FusedActionsObserver * __phase_observer = new FusedActionsObserver(); \
+   for ( FusedActions * __fuser__ : {static_cast<FusedActions *>(LoopFuser::getInstance()), static_cast<FusedActions *>(__phase_observer)}) { \
+      __fuser__->stopRecording(); \
```

Comment:
Should this be startRecording?

Reply:
no, with non-gpu, non CARE_DEBUG builds we default to executing loops as we encounter them (so we force recording to be off).